cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java

```diff
@@ -0,0 +1,234 @@
+                             Id.Namespace namespace,
+                             Map<String, String> runtimeArguments,
+                             @Nullable MetricsContext metricsContext) {
+    this.instantiator = instantiator;
+    this.txClient = txClient;
+    this.namespace = namespace;
+    this.runtimeArguments = runtimeArguments;
+    this.metricsContext = metricsContext;
+    this.extraTxAwares = Sets.newIdentityHashSet();
+  }
+
+  @Override
+  public final <T extends Dataset> T getDataset(String name)
+    throws DatasetInstantiationException {
+    return getDataset(name, DatasetDefinition.NO_ARGUMENTS);
+  }
+
+  @Override
+  public final <T extends Dataset> T getDataset(String name, Map<String, String> arguments)
+    throws DatasetInstantiationException {
+    return getDataset(name, arguments, false);
+  }
+
+  /**
+   * Instantiate a dataset, allowing to bypass the cache. This means that the dataset will not be added to
+   * the in-progress transactions, and it will also not be closed when the cache is closed.
+   * @param name the name of the dataset
+   * @param arguments arguments for the dataset
+   * @param bypass whether to bypass the cache
+   * @param <T> the type of the dataset
+   */
+  public final <T extends Dataset> T getDataset(String name, Map<String, String> arguments, boolean bypass)
+    throws DatasetInstantiationException {
+    // apply actual runtime arguments on top of the context's runtime arguments for this dataset
+    Map<String, String> dsArguments =
+      RuntimeArguments.extractScope(Scope.DATASET, name, runtimeArguments);
+    dsArguments.putAll(arguments);
+    return getDataset(new DatasetCacheKey(name, dsArguments), bypass);
+  }
+
+  /**
+   * To be implemented by subclasses.
+   *
+   * @param bypass if true, bypass the dataset cache, and do not add this to the transaction.
+   */
+  protected abstract <T extends Dataset> T getDataset(DatasetCacheKey key, boolean bypass)
+    throws DatasetInstantiationException;
+
+  /**
+   * Return a new transaction context. Any transaction-aware datasets obtained via
+   * (@link #getDataset()) will be added to this transaction context and thus participate
+   * in its transaction. These datasets can also be retrieved using {@link #getTransactionAwares()}.
+   *
+   * @return a new transactiopn context
+   */
+  public abstract TransactionContext newTransactionContext();
+
+  /**
+   * Dismiss the current transaction context. This releases the references to the context's
+   * transaction-aware datasets so that they can be collected by the garbage collector (if no one
+   * else is holding a reference to them).
+   */
+  public abstract void dismissTransactionContext();
+
+  @Override
+  public TransactionContext get() {
+    return newTransactionContext();
+  }
+
+  /**
+   * @return the transaction-aware datasets that participate in the current transaction.
+   */
+  public abstract Iterable<TransactionAware> getTransactionAwares();
```

Comment:
Actually it returns all tx-awares that are in the current transaction context. Or nothing if no tx context has been created. I updated the javadocs.